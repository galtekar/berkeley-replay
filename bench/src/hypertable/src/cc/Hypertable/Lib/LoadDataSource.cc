/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <cerrno>
#include <cctype>
#include <cstring>

#include <boost/algorithm/string.hpp>
#include <boost/shared_array.hpp>

extern "C" {
#include <strings.h>
#include <time.h>
}

#include "Common/ByteOrder.h"
#include "Common/DynamicBuffer.h"
#include "Common/Error.h"

#include "Key.h"

#include "LoadDataSource.h"

using namespace Hypertable;
using namespace std;

/**
 *
 */
LoadDataSource::LoadDataSource(std::string fname, std::string row_key_column, std::string timestamp_column) : m_fname(fname), m_cur_line(0), m_line_buffer(0), m_hyperformat(false), m_leading_timestamps(false), m_row_index(0), m_timestamp_index(-1), m_timestamp(0) {
   string line, column_name;
   char *base, *ptr;
   int index = 0;

   VK_MARK_FILE_BY_NAME(fname.c_str(), VK_PLANE_DATA);
   m_fin.open(fname.c_str());

   if (!getline(m_fin, line))
      return;

   base = (char *)line.c_str();
   //VK_MARK_DATA_REGION((ulong)base, strlen(base)+1);

   if (*base == '#') {
      base++;
      while (isspace(*base))
         base++;
   }
   while ((ptr = strchr(base, '\t')) != 0) {
      *ptr++ = 0;
      m_column_names.push_back(base);
      column_name = std::string(base);
      if (row_key_column != "" && row_key_column == column_name)
         m_row_index = index;
      if (timestamp_column != "" && timestamp_column == column_name)
         m_timestamp_index = index;
      base = ptr;
      index++;
   }
   m_column_names.push_back(base);

   if (m_column_names.size() == 3 || m_column_names.size() == 4) {
      size_t i=m_column_names.size()-3;
      if (m_column_names[i] == "rowkey" &&
            m_column_names[i+1] == "columnkey" &&
            m_column_names[i+2] == "value") {
         if (i == 0 || m_column_names[0] == "timestamp") {
            m_hyperformat = true;
            m_leading_timestamps = (i==1);
         }
      }
   }

   m_next_value = m_column_names.size();
   m_limit = 0;

   if (!m_hyperformat && m_column_names.size() < 2)
      throw Exception(Error::HQL_BAD_LOAD_FILE_FORMAT, "No columns specified in load file");

   m_cur_line = 1;

   return;
}


/**
 * 
 */
bool LoadDataSource::next(uint32_t *type_flagp, uint64_t *timestampp, KeySpec *keyp, uint8_t **valuep, uint32_t *value_lenp, uint32_t *consumedp) {
   string line;
   char *base, *ptr, *colon, *endptr;

   if (type_flagp)
      *type_flagp = FLAG_INSERT;

   if (consumedp)
      *consumedp = 0;

   if (m_hyperformat) {

      while (getline(m_fin, line)) {
         m_cur_line++;

         if (consumedp)
            *consumedp += line.length() + 1;

         m_line_buffer.clear();

         m_line_buffer.add(line.c_str(), strlen(line.c_str())+1);

         base = (char *)m_line_buffer.buf;

         /**
          *  Get timestamp
          */
         if (m_leading_timestamps) {
            if ((ptr = strchr(base, '\t')) == 0) {
               cerr << "error: too few fields on line " << m_cur_line << endl;
               continue;
            }
            *ptr++ = 0;
            *timestampp = strtoll(base, &endptr, 10);
            if (*endptr != 0) {
               cerr << "error: invalid timestamp (" << base << ") on line " << m_cur_line << endl;
               continue;
            }
            base = ptr;
         }
         else
            *timestampp = 0;

         /**
          * Get row key
          */
         if ((ptr = strchr(base, '\t')) == 0) {
            cerr << "error: too few fields on line " << m_cur_line << endl;
            continue;
         }
         keyp->row = base;
         keyp->row_len = ptr - base;
         *ptr++ = 0;
         base = ptr;

         /**
          * Get column family and qualifier
          */
         if ((ptr = strchr(base, '\t')) == 0) {
            cerr << "error: too few fields on line " << m_cur_line << endl;
            continue;
         }
         *ptr = 0;

         if ((colon = strchr(base, ':')) != 0) {
            *colon++ = 0;
            keyp->column_family = base;
            if (colon < ptr) {
               keyp->column_qualifier = colon;
               keyp->column_qualifier_len = ptr - colon;
            }
            else {
               keyp->column_qualifier = 0;
               keyp->column_qualifier_len = 0;
            }
         }
         else {
            keyp->column_qualifier = 0;
            keyp->column_qualifier_len = 0;
         }
         keyp->column_family = base;
         ptr++;

         /**
          * Get value
          */
         *valuep = (uint8_t *)ptr;
         *value_lenp = strlen(ptr);

         return true;
      }
   }
   else {

      while (m_next_value == (size_t)m_timestamp_index || m_next_value == (size_t)m_row_index)
         m_next_value++;

      if (m_next_value > 0 && m_next_value < (size_t)m_limit) {
         keyp->row = m_values[m_row_index];
         keyp->row_len = m_cur_row_length;
         keyp->column_family = m_column_names[m_next_value].c_str();
         *timestampp = m_timestamp;
         // clear these, just in case they were set by the client
         if (keyp->column_qualifier || keyp->column_qualifier_len) {
            keyp->column_qualifier = 0;
            keyp->column_qualifier_len = 0;
         }
         if (m_values[m_next_value] == 0) {
            *valuep = 0;
            *value_lenp = 0;
         }
         else {
            *valuep = (uint8_t *)m_values[m_next_value];
            *value_lenp = strlen(m_values[m_next_value]);
         }
         m_next_value++;
         return true;
      }

      while (getline(m_fin, line)) {
         m_cur_line++;

         if (consumedp)
            *consumedp += line.length() + 1;

         m_line_buffer.clear();
         m_line_buffer.add(line.c_str(), strlen(line.c_str())+1);
         m_values.clear();

         base = (char *)m_line_buffer.buf;

         while ((ptr = strchr(base, '\t')) != 0) {
            *ptr++ = 0;
            if (strlen(base) == 0 || !strcmp(base, "NULL") || !strcmp(base, "\\N"))
               m_values.push_back(0);
            else
               m_values.push_back(base);
            base = ptr;
         }
         if (strlen(base) == 0 || !strcmp(base, "NULL") || !strcmp(base, "\\N"))
            m_values.push_back(0);
         else
            m_values.push_back(base);
         m_limit = std::min(m_values.size(), m_column_names.size());

         if (m_timestamp_index >= 0) {
            struct tm tm;
            time_t t;

            if (m_values.size() <= (size_t)m_timestamp_index) {
               cerr << "warn: timestamp field not found on line " << m_cur_line << ", skipping..." << endl;
               continue;
            }	

            if (!parse_date_format(m_values[m_timestamp_index], &tm)) {
               cerr << "warn: invalid timestamp format on line " << m_cur_line << ", skipping..." << endl;
               continue;
            }

            if ((t = timegm(&tm)) == (time_t)-1) {
               cerr << "warn: invalid timestamp format on line " << m_cur_line << ", skipping..." << endl;
               continue;
            }

            m_timestamp = (uint64_t)t * 1000000000LL;
         }

         if (m_values[m_row_index] == 0) {
            cerr << "error: NULL row key on line " << m_cur_line << endl;
            continue;
         }

         m_next_value = 0;
         while (m_next_value == (size_t)m_timestamp_index || m_next_value == (size_t)m_row_index)
            m_next_value++;

         m_cur_row_length = strlen(m_values[m_row_index]);

         keyp->row = m_values[m_row_index];
         keyp->row_len = m_cur_row_length;
         keyp->column_family = m_column_names[m_next_value].c_str();
         *timestampp = m_timestamp;
         if (keyp->column_qualifier || keyp->column_qualifier_len) {
            keyp->column_qualifier = 0;
            keyp->column_qualifier_len = 0;
         }
         if (m_values[m_next_value] == 0) {
            *valuep = 0;
            *value_lenp = 0;
         }
         else {
            *valuep = (uint8_t *)m_values[m_next_value];
            *value_lenp = strlen(m_values[m_next_value]);
         }
         m_next_value++;

         return true;
      }

   }

   return false;
}



bool LoadDataSource::parse_date_format(const char *str, struct tm *tm) {
   int ival;
   const char *ptr = str;
   char *end_ptr;

   /**
    * year
    */
   if ((ival = strtol(ptr, &end_ptr, 10)) == 0 || (end_ptr - ptr) != 4 || *end_ptr != '-')
      return false;
   tm->tm_year = ival - 1900;

   ptr = end_ptr + 1;

   /**
    * month
    */
   if ((ival = strtol(ptr, &end_ptr, 10)) == 0 || (end_ptr - ptr) != 2 || *end_ptr != '-')
      return false;
   tm->tm_mon = ival - 1;

   ptr = end_ptr + 1;

   /**
    * day
    */
   if ((ival = strtol(ptr, &end_ptr, 10)) == 0 || (end_ptr - ptr) != 2 || *end_ptr != ' ')
      return false;
   tm->tm_mday = ival;

   ptr = end_ptr + 1;

   /**
    * hour
    */
   ival = strtol(ptr, &end_ptr, 10);
   if ((end_ptr - ptr) != 2 || *end_ptr != ':')
      return false;
   tm->tm_hour = ival;

   ptr = end_ptr + 1;

   /**
    * minute
    */
   ival = strtol(ptr, &end_ptr, 10);
   if ((end_ptr - ptr) != 2 || *end_ptr != ':')
      return false;
   tm->tm_min = ival;

   ptr = end_ptr + 1;

   /**
    * second
    */
   ival = strtol(ptr, &end_ptr, 10);
   if ((end_ptr - ptr) != 2)
      return false;
   tm->tm_sec = ival;

   tm->tm_gmtoff = 0;
   tm->tm_zone = "GMT";

   return true;
}
