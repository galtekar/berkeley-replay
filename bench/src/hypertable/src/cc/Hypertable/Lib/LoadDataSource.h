/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef HYPERTABLE_LOADDATASOURCE_H
#define HYPERTABLE_LOADDATASOURCE_H

#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "Common/ByteString.h"
#include "Common/DynamicBuffer.h"

#include "DataSource.h"
#include <vk.h>

using namespace Hypertable;

namespace Hypertable {

  class LoadDataSource {

  public:
    LoadDataSource(std::string fname, std::string row_key_column, std::string timestamp_column);
    virtual ~LoadDataSource() { VK_MARK_FILE_BY_NAME(m_fname.c_str(), VK_PLANE_DATA); return; }
    bool has_timestamps() { return m_leading_timestamps || (m_timestamp_index != -1); }
    virtual bool next(uint32_t *type_flagp, uint64_t *timestampp, KeySpec *keyp, uint8_t **valuep, uint32_t *value_lenp, uint32_t *consumedp);

  private:

    bool parse_date_format(const char *str, struct tm *tm);

    std::vector<std::string> m_column_names;
    std::vector<const char *> m_values;
    size_t m_next_value;
    std::ifstream m_fin;
    std::string m_fname;
    long m_cur_line;
    DynamicBuffer m_line_buffer;
    bool m_hyperformat;
    bool m_leading_timestamps;
    size_t m_cur_row_length;
    int m_row_index;
    int m_timestamp_index;
    uint64_t m_timestamp;
    int m_limit;
  };

}

#endif // HYPERTABLE_LOADDATASOURCE_H
