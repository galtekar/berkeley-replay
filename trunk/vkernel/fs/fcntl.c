#include "vkernel/public.h"
#include "private.h"

void 
Fcntl_SetCloseOnExec(unsigned int fd, int flag)
{
	struct FilesStruct *files = current->files;
	struct FdTableStruct *fdt;

	Files_Lock(files);

	fdt = files->fdt;
	if (flag)
		FD_SET(fd, fdt->closeOnExec);
	else
		FD_CLR(fd, fdt->closeOnExec);

   Files_Unlock(files);
}

static int
FcntlGetCloseOnExec(unsigned int fd)
{
	struct FilesStruct *files = current->files;
	struct FdTableStruct *fdt;
	int res;

   Files_Lock(files);

	fdt = files->fdt;
	res = FD_ISSET(fd, fdt->closeOnExec);

   Files_Unlock(files);

	return res;
}

/*
 * The file flags determines the behavior of file objects with
 * respect to file operations.
 * For example, the flags determine issues such as, will the operation
 * block? Or will we get asynchronous notification if data
 * becomes available through a file object?
 *
 * The job of interpreting and implementing these parameters are
 * delegated to the device VFS calls. The Pipe and Linux device, for example,
 * simply invokes the Linux setfl syscall on its backing fd 
 * and lets the kernel
 * handle the flag changes and the task of providing appropriate
 * VFS operation return values and signals.
 */
static int
FcntlFlagOp(struct FileStruct *filp, int isSet, ulong flags)
{
   int oldFlags;

   ASSERT(File_IsLocked(filp));

   oldFlags = filp->flags;

   if (isSet) {
      filp->flags = flags;
   }

   return oldFlags;
}

static int
FcntlFlagsGet(struct FileStruct *filp)
{
   int flags;

   File_Lock(filp);

   flags = FcntlFlagOp(filp, 0, 0);

   File_Unlock(filp);

   return flags;
}


int
Fcntl_FlagsOp(struct FileStruct *filp, ulong flags, int op)
{
   int err;
   ulong newFlags;

   /* required for strict SunOS emulation */
   if (O_NONBLOCK != O_NDELAY)
      if (flags & O_NDELAY)
         flags |= O_NONBLOCK;

   /* Ensure that accesses to the file flags are deterministic. */
   File_Lock(filp);

#define SETFL_MASK (O_APPEND | O_NONBLOCK | O_NDELAY | FASYNC | O_DIRECT | O_NOATIME)
   newFlags = filp->flags;
   switch (op) {
   case FCNTL_OP_SET:
      newFlags &= ~SETFL_MASK;
      newFlags |= (flags & SETFL_MASK);
      break;
   case FCNTL_OP_OR:
      newFlags |= (flags & SETFL_MASK);
      break;
   case FCNTL_OP_AND:
      newFlags &= (flags & SETFL_MASK);
      break;
   default:
      ASSERT(0);
      break;
   }

   if (!filp->f_op || !filp->f_op->setfl) {
      err = -EINVAL;
      goto out;
   }

   err = filp->f_op->setfl(filp, newFlags);

   /* XXX: why do we need to shadow the file flags? */
   if (!err) {
      FcntlFlagOp(filp, 1, newFlags);
   }

out:
   File_Unlock(filp);
   return err;
}

#if 0
static INLINE int 
flock_translate_cmd(int cmd) {
	if (cmd & LOCK_MAND)
		return cmd & (LOCK_MAND | LOCK_RW);
	switch (cmd) {
	case LOCK_SH:
		return F_RDLCK;
	case LOCK_EX:
		return F_WRLCK;
	case LOCK_UN:
		return F_UNLCK;
	}
	return -EINVAL;
}

/* Fill in a file_lock structure with an appropriate FLOCK lock. */
static int 
flock_make_lock(struct flock *fl, unsigned int cmd)
{
	int type = flock_translate_cmd(cmd);
	if (type < 0)
		return type;

   DEBUG_MSG(5, "cmd=%d type=%d\n", cmd, type);
	
	fl->l_type = type;
   fl->l_whence = SEEK_SET;
   fl->l_start = 0;
	fl->l_len = 0; /* 0 <==> OFFSET_MAX (see flock_to_posix_lock in Linux) */
	fl->l_pid = current->tgid;
	
	return 0;
}
#endif

/* 
 * sys_flock and the fcntl(F_GETLK/F_SETLK) are advisory locks
 * and are implemented identically in Linux.
 */
long 
sys_flock(unsigned int fd, unsigned int cmd)
{
   struct FileStruct *filp;
#if 0
   struct flock lock;
   int can_sleep, unlock;
#endif
   int err = -EBADF;

   filp = File_Get(fd);
   if (!filp) {
      goto out;
   }

#if 0
   can_sleep = !(cmd & LOCK_NB);
   cmd &= ~LOCK_NB;
   unlock = (cmd == LOCK_UN);

   if (!unlock && !(cmd & LOCK_MAND) && !(filp->accMode & 3)) {
      D__;
      goto out_putf;
   }

   err = flock_make_lock(&lock, cmd);
   if (err) {
      goto out_putf;
      D__;
   }
#endif

   if (filp->f_op && filp->f_op->lock) {
      err = filp->f_op->flock(filp, cmd);
      D__;
   } else {
      //error = flock_lock_file_wait(filp, lock);
      ASSERT_UNIMPLEMENTED(0);
   }

   File_Put(filp);
out:
   return err;
}

static long
FcntlGetLock(struct FileStruct *filp, struct flock __user * l) 
{
	struct flock flock;
	int err;

	err = -EFAULT;
	if (Task_CopyFromUser(&flock, l, sizeof(flock)))
		goto out;


   err = -EINVAL;
	if (filp->f_op && filp->f_op->lock) {
		err = filp->f_op->lock(filp, F_GETLK, &flock);
   } else {
      goto out;
   }

	err = -EFAULT;
	if (!Task_CopyToUser(l, &flock, sizeof(flock)))
		err = 0;
out:
	return err;
}

static long
FcntlSetLock(UNUSED uint vfd, struct FileStruct *filp, uint cmd, 
             const struct flock __user *l)
{
	struct flock flock;
	int err;

	err = -EFAULT;
	if (Task_CopyFromUser(&flock, l, sizeof(flock)))
		goto out;


   err = -EINVAL;
	if (filp->f_op && filp->f_op->lock) {
		err = filp->f_op->lock(filp, cmd, &flock);
   } else {
      goto out;
   }

out:
	return err;
}

static long
FcntlGetLock64(struct FileStruct *filp, struct flock64 __user * l) 
{
	struct flock64 flock;
	int err;

	err = -EFAULT;
	if (Task_CopyFromUser(&flock, l, sizeof(flock)))
		goto out;


   err = -EINVAL;
	if (filp->f_op && filp->f_op->lock) {
		err = filp->f_op->lock(filp, F_GETLK64, &flock);
   } else {
      goto out;
   }

	err = -EFAULT;
	if (!Task_CopyToUser(l, &flock, sizeof(flock)))
		err = 0;
out:
	return err;
}

static long
FcntlSetLock64(UNUSED uint vfd, struct FileStruct *filp, uint cmd, 
             const struct flock64 __user *l)
{
	struct flock64 flock;
	int err;

	err = -EFAULT;
	if (Task_CopyFromUser(&flock, l, sizeof(flock)))
		goto out;


   err = -EINVAL;
	if (filp->f_op && filp->f_op->lock) {
		err = filp->f_op->lock(filp, cmd, &flock);
   } else {
      goto out;
   }

out:
	return err;
}


static long
FcntlDo(uint vfd, uint cmd, ulong arg, struct FileStruct* file)
{
   long err = -EINVAL;

   switch (cmd) {
   case F_DUPFD:
      /* XXX: call the internal dup function, since we already
       * have a handle to FileStruct. */
      err = sys_dup(vfd);
      break;
   case F_GETFD:
      err = FcntlGetCloseOnExec(vfd) ? FD_CLOEXEC : 0;
      break;
   case F_SETFD:
      err = 0;
      Fcntl_SetCloseOnExec(vfd, arg & FD_CLOEXEC);
      break;
   case F_GETFL:
      err = FcntlFlagsGet(file);
      break;
   case F_SETFL:
      err = Fcntl_FlagsOp(file, arg, FCNTL_OP_SET);
      break;
   case F_GETLK:
      /* Determines whether lock is already held and if so returns info
       * about existing lock. */
      err = FcntlGetLock(file, (struct flock __user *) arg);
      break;
   case F_SETLK:
      /* Acquire a lock. If cannot be acquired, then returns -- does not
       * block. */
   case F_SETLKW:
      /* Like F_SETLK, but blocks if lock cannot be acquired. */
      err = FcntlSetLock(vfd, file, cmd, (struct flock __user *) arg);
      break;

   default:
      ASSERT_UNIMPLEMENTED(0);
      break;
   }

   return err;
}


long
sys_fcntl64(uint vfd, uint cmd, ulong arg)
{
   struct FileStruct *file;
   SyscallRet err;

   err = -EBADF;
   file = File_Get(vfd);
   if (!file) {
      goto out;
   }

   DEBUG_MSG(5, "vfd=%d cmd=0x%x (%d)\n", vfd, cmd, cmd);

   switch (cmd) {
      /* MySql uses these, apparently. */
   case F_GETLK64:
      /* Determines whether lock is already held and if so returns info
       * about existing lock. */
      err = FcntlGetLock64(file, (struct flock64 __user *) arg);
      break;
   case F_SETLK64:
      /* Acquire a lock. If cannot be acquired, then returns -- does not
       * block. */
   case F_SETLKW64:
      /* Like F_SETLK, but blocks if lock cannot be acquired. */
      err = FcntlSetLock64(vfd, file, cmd, (struct flock64 __user *) arg);
      break;
   default:
      err = FcntlDo(vfd, cmd, arg, file);
      break;
   }

   File_Put(file);
out:
   return err;
}

long
sys_fcntl(uint vfd, uint cmd, ulong arg)
{
   return sys_fcntl64(vfd, cmd, arg);
}
