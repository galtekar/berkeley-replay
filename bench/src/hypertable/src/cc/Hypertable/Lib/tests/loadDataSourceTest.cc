/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <cstdlib>
#include <iostream>

extern "C" {
#include <errno.h>
#include <fcntl.h>
}


#include "Hypertable/Lib/KeySpec.h"
#include "Hypertable/Lib/LoadDataSource.h"

using namespace Hypertable;
using namespace std;

int main(int argc, char **argv) {
  LoadDataSource  *lds;
  uint64_t timestamp;
  KeySpec key;
  uint8_t *value;
  uint32_t value_len;
  int fd;

  if ((fd = open("loadDataSourceTest.output", O_WRONLY|O_CREAT|O_TRUNC, 0644)) < 0) {
    perror("open");
    return 1;
  }

  close(2);
  dup(fd);

  lds = new LoadDataSource("loadDataSourceTest.dat", "", "");

  while (lds->next(0, &timestamp, &key, &value, &value_len, 0)) {
    cerr << "row=" << (const char *)key.row << " column_family=" << key.column_family;
    if (key.column_qualifier_len > 0)
      cerr << " column_qualifier=" << (const char *)key.column_qualifier;
    cerr << " value=" << (const char *)value << endl;
  }

  delete lds;

  close(2);
  dup(1);

  if (system("diff loadDataSourceTest.output loadDataSourceTest.golden") != 0)
    return 1;

  return 0;
}
