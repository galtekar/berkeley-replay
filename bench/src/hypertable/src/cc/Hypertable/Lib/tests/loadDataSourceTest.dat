timestamp	rowkey	columnkey	value
0	abenteric	foo:http://www.foo.com/	How now brown cow
123456789	barothermohygrograph	bar	Herbie Hancock
98787652	Batonga	foo:	Head Hunters
0
abracadabra	Batonga	foo:	Head Hunters
123456789	barothermohygrogra
123456789	barothermohygrograph	ba
0	carbonification	sherlock:holmes	this\tis\ta\ttab\tdelimited\tvalue
