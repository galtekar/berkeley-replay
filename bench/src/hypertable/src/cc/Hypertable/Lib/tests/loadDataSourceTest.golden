row=abenteric column_family=foo column_qualifier=http://www.foo.com/ value=How now brown cow
row=barothermohygrograph column_family=bar value=Herbie Hancock
row=Batonga column_family=foo value=Head Hunters
error: too few fields on line 5
error: invalid timestamp (abracadabra) on line 6
error: too few fields on line 7
error: too few fields on line 8
row=carbonification column_family=sherlock column_qualifier=holmes value=this\tis\ta\ttab\tdelimited\tvalue
